### ISSUE TITLE
avoid the need for `bundle exec` prefix

### ISSUE THREAD

**Issue by olistik** [hanami/hanami#676]
## Expected behavior

```shell
$ hanami routes
# ...
$ hanami server
```

## Current behavior

```shell
$ bundle exec hanami routes
# ...
$ bundle exec hanami server
```

## The Rails-way

```bash
$ which rails
/Users/olistik/.rvm/gems/ruby-2.3.1/bin/rails
```

In `/Users/olistik/.rvm/gems/ruby-2.3.1/bin/rails`:

```ruby
load Gem.bin_path('railties', 'rails', version)
# /Users/olistik/.rvm/gems/ruby-2.3.1/gems/railties-5.0.0.1/exe/rails
```
## railties gem

Within `/Users/olistik/.rvm/gems/ruby-2.3.1/gems/railties-5.0.0.1`:

```ruby
# ./exe/rails
require "rails/cli"
```

```ruby
# ./lib/rails/cli.rb
require 'rails/app_loader'
# ...
Rails::AppLoader.exec_app
```

```ruby
# ./lib/rails/app_loader.rb
# searches for the presence of the script `bin/rails` within the current directory `Dir.pwd`
# *if* there's such file, then loads the corresponding `config/boot.rb`
  Object.const_set(:APP_PATH, File.expand_path('config/application', Dir.pwd))
  require File.expand_path('../boot', APP_PATH)

    # ./lib/rails/generators/rails/app/templates/config/boot.rb
    ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../Gemfile', __dir__)
    require 'bundler/setup' # Set up $LOAD_PATH according to the Gemfile.
    # subsequent invocations of `require` will respect the newly setup $LOAD_PATH

  # with $LOAD_PATH set accordingly to the application's Gemfile, requires the CLI logic
  require 'rails/commands'
  
# *else* the command is invoked likely to create a new application so give the control back to the version provided
  # ...
```

## TL;DR

It boils down to detecting the context of the caller (creating a new Hanami app or invoking a command within an already existing application) and then requiring ’bundler/setup’ in the latter case.

This way we can avoid prepending bundle exec when invoking each `hanami` command.

**Comment by cllns** [comment]
Thanks for this issue. The way I get around this is alias `bundle` to `b`, so I can do `b e` which is now just how I prepend all my ruby commands (I also have `be` as `bundle exec`, for typos 😇)

Overall, this change would improve the immediate user experience, since it means developers have less to worry about and learn. In the long run though, it obscures the importance and meaning of `bundle exec`. We should be enabling and encouraging people to learn more about Ruby, not hiding complexity from them by hacking the Ruby ecosystem.

What do others think?

**Comment by b264** [comment]
Encouraging learning more about the Ruby ecosystem is great, but also anything making it more difficult for beginners to start is a net loss for all of us, in my opinion.  Also, the trend in CS is to abstract away more mundane details as time goes by and start focusing more on the core issues making your software different from other software.  So, while I _highly_ appreciate the point made above, I'd have to vote for abstracting it away and avoid prepending `bundle exec` when invoking each `hanami` command.

**Comment by jodosha** [comment]
It would be nice to have.

@olistik What happens when someone still types: `bundle exec rails ...`?

@cllns I'm not sure if your pro or against this feature.

**Comment by olistik** [comment]
@jodosha 

> What happens when someone still types: bundle exec rails ...?

Just Bundler doing the same thing twice, that is slowing the startup time of the command.
That would be useless.

**Comment by jodosha** [comment]
@olistik Cool, just wanted to make sure it was working. Thanks.

**Comment by cllns** [comment]
@jodosha I'm weakly against it. If others want it, that's fine :)

**Comment by TiteiKo** [comment]
I agree with @cllns arguments, but I'm a bit more than weakly against it..

There's 3 types of people at hand here:

* **Ruby total beginners** who are reading the docs, and who most likely copy/paste the given commands. For them, I think we should add a quick note on why Bundler, its role and how to make an alias for it.
* **Hanami beginners** who already know bundler, have their alias, they use it everyday.
* **Hanami everyday user**: Same as above.

---

The expected behaviour for a gem executable is the one being in place atm:
* you can use the commands directly, and it will use the latest-installed version of the gems executable
* to be safe and use the version of your Gemfile.lock, use `bundle exec`

Hacking this not only slows things for people doing the logical thing, but creates an(other) exception to the expected behaviour...

**Comment by olistik** [comment]
@TiteiKo 

> Hacking this not only slows things for people doing the logical thing, but creates an(other) exception to the expected behaviour...

This looks a fair point to me but I'm weakly (non totally :-)) against it because having a command that adjusts itself according to the context of the caller looks legit to me.

The message to the users could simply be

> Every `hanami` command will run according to the dependencies of the project's Gemfile.

This could be just the expected behavior of a Ruby script with the added value of having a less verbose and central entry point to the project's CLI interactions.
I see a wide set of beginners caring about the raison d'être of the commands given to them.

`bundle exec hanami` is three things to them:

1. `bundle exec`: "what for?"
2. `hanami`: "ok, this one gives commands"
3. 1 + 2: "why are they needed together?"

I'm not arguing that understanding `bundle exec` is a strict requirement in the path of a Ruby developer but instead that having to deal with more than one concept at the same time, at least for novice users, can be a bit challenging.

Having said that, I can feel the hack-ish sensation of having added an exception.

**Comment by TrumpClone** [comment]
without this proposal: we have two options (thx @TiteiKo) => 
#1) you can use the commands directly, and it will use the latest-installed version of the gems executable
#2) to be safe and use the version of your Gemfile.lock, use bundle exec

but with this proposal: only #1 (with versions in Gemfile file)

is it good?

**Comment by b264** [comment]
I think it's good because it can avoid things not working because the wrong (not Gemfile) versions are being run.  Think of someone learning Hanami and it's not working because of this.  Sure, when we know what's going on things are easy to fix, but I'm thinking of folks who are still a novice.  When I started, `bundle exec` was super-confusing.  The real question here is should Hanami be beginner-friendly, I think.  I'm not sure anyone could disagree that requiring `bundle exec` is "technically better".

**Comment by erwanthomas** [comment]
Bundler binstubs are helpful here: `bundle exec hanami` becomes `bin/hanami`.

```bash
$ bundle binstubs hanami
$ cat bin/hanami
```
```ruby
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# This file was generated by Bundler.
#
# The application 'hanami' is installed as part of a gem, and
# this file is here to facilitate running it.
#

require "pathname"
ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../../Gemfile",
  Pathname.new(__FILE__).realpath)

require "rubygems"
require "bundler/setup"

load Gem.bin_path("hanami", "hanami")
```

Advantages:
* calls the safe, bundled executable
* requires only 4 more keystrokes
* no magic required from the Hanami executable (actually, no change at all)
* nice to the user:
  * uses Bundler standards (same technique can be used for any gem)
  * can teach beginners about what `bundle exec` actually does
  * already known by advanced users

**Comment by sandelius** [comment]
I have no problems running `be` (alias) before commands, that makes it easier for my brain to know which context I'm currently in.

**Comment by cllns** [comment]
In my opinion, `binstubs` is a good compromise. What do others think?

**Comment by b264** [comment]
I'd go for binstubs

**Comment by olistik** [comment]
FWIW I'd go for `binstubs` as well :-)
@jodosha should I close the issue?

**Comment by gizotti** [comment]
👍  to `binstubs`

**Comment by cllns** [comment]
@olistik You can either close this issue and open a new one about `binstubs` (and reference this issue there!) or just edit this title and first post to reflect that binstubs is the approach we're going for :)

**Comment by olistik** [comment]
@cllns I'm going to create a new issue. What would the goal be? Just to suggest the use of bin stubs in the README? I guess there's no additional code required.

**Comment by cllns** [comment]
Rather than instructing users to run `bundle binstubs hanami`, I was thinking we could generate a `bin/hanami` file inside new hanami projects (like Rails does, with `bin/rails`). 

What do others think of that?

**Comment by mereghost** [comment]
Closing this then. @olistik reference this issue when you open the new issue and the discussion can continue over there. =D

One way we could enforce this is generate a `.bundler/config` as follows:
```
---
BUNDLE_BIN: "bin"' 
```
so bundler will always generate binstubs when the user calls bundle install.
